velox/functions/prestosql/Sequence.cpp

```diff
@@ -0,0 +1,94 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/expression/VectorFunction.h"
+#include "velox/vector/ConstantVector.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+// See documentation at https://prestodb.io/docs/current/functions/array.html
+class SequenceFunction : public exec::VectorFunction {
+ public:
+  static std::vector<std::shared_ptr<exec::FunctionSignature>> signatures() {
+    std::vector<std::shared_ptr<exec::FunctionSignature>> signatures;
+
+    // integer, integer -> array(integer)
+    signatures.emplace_back(exec::FunctionSignatureBuilder()
+                                .returnType("array(integer)")
+                                .argumentType("integer")
+                                .argumentType("integer")
+                                .build());
+
+    // todo integer, integer, integer -> array(integer)
+    signatures.emplace_back(exec::FunctionSignatureBuilder()
+                                .returnType("array(integer)")
+                                .argumentType("integer")
+                                .argumentType("integer")
+                                .argumentType("integer")
+                                .build());
+
+    // date, date -> array(date)
+    return signatures;
+  }
+
+ private:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    auto pool = context.pool();
+    auto startVector = args[0]->asFlatVector<int32_t>();
+    auto endVector = args[1]->asFlatVector<int32_t>();
+    context.ensureWritable(rows, outputType, result);
+    result->clearNulls(rows);
+    auto arrayResult = result->as<ArrayVector>();
+    auto elementsResult = arrayResult->elements();
+
+    auto sizes = arrayResult->mutableSizes(rows.size());
+    auto rawSizes = sizes->asMutable<int32_t>();
+    auto offsets = arrayResult->mutableOffsets(rows.size());
+    auto rawOffsets = offsets->asMutable<int32_t>();
+
+    auto baseOffset = elementsResult->size();
+    vector_size_t offset = baseOffset;
+
+    rows.applyToSelected([&](vector_size_t row) {
+      auto start = startVector->valueAt(row);
+      auto end = endVector->valueAt(row);
+      rawOffsets[row] = offset;
+      rawSizes[row] = end - start + 1;
+      elementsResult->resize(offset + rawSizes[row]);
+      for (auto i = start; i <= end; i++) {
+        auto vector =
+            BaseVector::create<FlatVector<int32_t>>(BIGINT(), 1, pool);
+        auto flatVector = vector->asFlatVector<int32_t>();
+        flatVector->mutableRawValues()[0] = i;
+        elementsResult->copy(flatVector, offset++, row, 1);
```

Comment:
From what I understand here we create a 1-element vector for every element and then copy it (single item) to the target elements vector.
This would be very inefficient. We should write to elements directly.